code-challenge-android-kotlin/app/src/main/java/com/arctouch/codechallenge/ui/HomeActivity.kt

```diff
@@ -0,0 +1,105 @@
+package com.arctouch.codechallenge.ui
+
+import android.arch.lifecycle.Observer
+import android.arch.lifecycle.ViewModelProviders
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.view.View
+import com.arctouch.codechallenge.viewmodel.HomeViewModel
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.widget.AbsListView
+import com.arctouch.codechallenge.R
+import com.arctouch.codechallenge.data.Cache
+import com.arctouch.codechallenge.home.HomeAdapter
+import com.arctouch.codechallenge.model.Genre
+import com.arctouch.codechallenge.model.Movie
+import com.arctouch.codechallenge.webservice.MoviesWS
+import kotlinx.android.synthetic.main.home_activity.*
+
+class HomeActivity : AppCompatActivity() {
+    private var isScrolling = false
+    private var currentItems = 0
+    private var scrolledOutItems = 0
+    private var totalItems = 0
+    private var upcomingMovies = ArrayList<Movie>()
+    private var pageNumber = 1L
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.home_activity)
+
+        val viewModel = ViewModelProviders.of(this).get(HomeViewModel::class.java)
+
+        viewModel.getGenres().observe(this, Observer<List<Genre>>{genres ->
+            viewModel.getUpcomingMovies(pageNumber).observe(this, Observer<List<Movie>>{ movies ->
+                movies?.let{
+                    upcomingMovies = ArrayList(movies)
+                    val layoutManager = LinearLayoutManager(this)
+                    val adapter = HomeAdapter(this, upcomingMovies, showMovieFragmentLayout,
+                    hideMovieFragmentLayout, showProgressBar, hideProgressBar)
+                    recyclerView.layoutManager = layoutManager
+                    recyclerView.adapter = adapter
+                    val scrollListener = getScrollListener(layoutManager, adapter)
+                    recyclerView.addOnScrollListener(scrollListener)
+                    progressBar.visibility = View.GONE
+            }
+
+                //TODO notify user if movies = null
+            })
+        })
+    }
+
+    var showMovieFragmentLayout: () -> Unit = {
```

Comment:
Why did you pass these variables as high-order functions and lambdas to the `HomeAdapter` constructor? Should `HomeAdapter` need to how exactly about showing/hiding fragment, showing/hiding progressbar?

Reply:
Let me know if I got your question wrong. I think a word is missing. "Should HomeAdapter need to **know** how exactly about showing... "

I passed them as high-order functions because I needed to call them from the activity, and I prefer using them instead of listeners (as in Java). But the reason for calling them from the activity is because I wanted the progressbar to show when each movie was selected, while API result was still being loaded, and for the layout which contained the fragment to show/hide. 

It seemed like it wasn't a good idea and like there was a better way to do it, but I couldn't come up with something better to solve the problem at the time. 

I don't think `HomeAdapter ` should know/access Activity code thought I don't know why.